sampler_test.go

```diff
@@ -575,3 +578,106 @@ func TestRemotelyControlledSampler_updateSamplerFromAdaptiveSampler(t *testing.T
 	_, ok = remoteSampler.sampler.(*rateLimitingSampler)
 	require.True(t, ok)
 }
+
+func TestAdaptiveSampler_lockRaceCondition(t *testing.T) {
+	agent, remoteSampler, _ := initAgent(t)
+	defer agent.Close()
+	remoteSampler.Close() // stop timer-based updates, we want to call them manually
+
+	numOperations := 100
+	strategiesA := generatePerOperationStrategies(numOperations, "a")
+	strategiesB := generatePerOperationStrategies(numOperations, "b")
+	adaptiveSampler, err := NewAdaptiveSampler(strategiesA, 200)
+	require.NoError(t, err)
+
+	// Overwrite the sampler with an adaptive sampler
+	remoteSampler.sampler = adaptiveSampler
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	wg.Add(3)
+
+	exit1 := make(chan struct{})
+	exit2 := make(chan struct{})
+	exit3 := make(chan struct{})
+
+	go func() {
+		defer wg.Done()
+		isSampled(t, exit1, remoteSampler, numOperations, "a")
+	}()
+	go func() {
+		defer wg.Done()
+		isSampled(t, exit2, remoteSampler, numOperations, "b")
+	}()
+	go func() {
+		defer wg.Done()
+		updateSampler(
+			exit3,
+			remoteSampler,
+			agent,
+			&sampling.SamplingStrategyResponse{OperationSampling: strategiesA},
+			&sampling.SamplingStrategyResponse{OperationSampling: strategiesB},
+		)
+	}()
+	time.Sleep(10 * time.Second)
```

Comment:
1. do we need to do that check through end-to-end testing via remote sampler?  Looks like we could do it only on the adaptive sampler.
  1. if we do the latter, what if you just run those routines a lot faster? I.e. why sleep 10us / 100us? Can we instead call Gosched() (https://golang.org/pkg/runtime/#Gosched) to yield from each goroutine, and have a small overall wait interval of say 10ms to terminate the test. If you run such test without your fix and it panics, then it's ok to keep it running permanently, 10ms isn't a big deal.